File: CharacterSheetCompanion/CharacterSheetCompanion/Services/SocketListeners/ActorListeners/ActorTraitsListener.swift
//
//  ActorBaseDataListener.swift
//  CharacterSheetCompanion
//
//  Created by Zachary Sanders on 8/17/21.
//

import Combine
import Foundation
import SocketIO

class ActorTraitsListener: SocketListener, ActorListener {
    let traitsPublisher: AnyPublisher<ActorTraitsModel?, Never>
    let socket: SocketIOClient

    private let traitsSubject = CurrentValueSubject<ActorTraitsModel?, Never>(nil)

    init(socket: SocketIOClient) {
        traitsPublisher = traitsSubject.eraseToAnyPublisher()
        self.socket = socket
    }

    func addSocketHandlers() {
        socket.on(SocketEvents.SERVER.ACTOR.SEND.SEND_ACTOR_TRAITS) { data, _ in
            do {
                try self.traitsSubject.send(SocketListenerUtility.parseSocketEventData(data))
            } catch let FoundryJSONError.errorMessage(errorMessage) {
                print(errorMessage)
            } catch {
                print(error)
            }
        }
    }

    func requestInitialActorData(actorId: String) {
        socket.emit(SocketEvents.IOS.ACTOR.REQUEST_ACTOR_TRAITS, actorId)
    }

    func isReady() -> Bool {
        true
    }
}

extension SocketEvents.IOS.ACTOR {
    static let REQUEST_ACTOR_TRAITS = "ios:requestActorTraits"
}

extension SocketEvents.SERVER.ACTOR.SEND {
    static let SEND_ACTOR_TRAITS = "server:sendActorTraits"
}

File: character-sheet-companion-server/src/listeners/actor/actorTraitsListener.ts
import { Server, Socket } from "socket.io";
import {
  getFoundrySocketFromRoom,
  getIosSocketFromRoom,
} from "../../utilities/SocketUtilities";
import { SOCKET_EVENTS } from "../../constants/events";
import { ActorTraits } from "../../interfaces/actor/ActorTraits";

export const actorTraitsListener = (io: Server, socket: Socket): void => {
  const requestActorTraits = (actorId: string) => {
    const foundrySocket = getFoundrySocketFromRoom(io, socket);
    if (foundrySocket) {
      foundrySocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.REQUEST.REQUEST_ACTOR_TRAITS,
        actorId,
        socket.id
      );
    }
  };

  const receiveActorTraits = (
    actorTraits: ActorTraits,
    iosSocketId: string
  ) => {
    const iosSocket = getIosSocketFromRoom(io, socket, iosSocketId);
    if (iosSocket) {
      iosSocket.emit(
        SOCKET_EVENTS.SERVER.ACTOR.SEND.SEND_ACTOR_TRAITS,
        JSON.stringify(actorTraits)
      );
    }
  };

  socket.on(SOCKET_EVENTS.IOS.ACTOR.REQUEST_ACTOR_TRAITS, requestActorTraits);
  socket.on(SOCKET_EVENTS.FOUNDRY.ACTOR.SEND_ACTOR_TRAITS, receiveActorTraits);
};
